File: config/initializers/kafka.rb
# frozen_string_literal: true

Rails.application.config.kafka = {
  brokers: ENV.fetch("KAFKA_BROKERS", "localhost:9092").split(","),
  client_id: "moviedb"
}

unless ENV.key?("KAFKA_BROKERS") || Rails.env.test?
  Rails.logger.warn("KAFKA_BROKERS not set; defaulting to localhost:9092")
end

File: lib/kafka.ts
import { Kafka } from 'kafkajs';

const kafka = new Kafka({
  clientId: 'moviedb',
  brokers: process.env.KAFKA_BROKERS?.split(',') || ['localhost:9092']
});

const producer = kafka.producer();

export async function publishAudit(
  action: 'created' | 'updated' | 'deleted',
  recordId: string,
  before: Record<string, unknown> | null,
  after: Record<string, unknown> | null
): Promise<void> {
  try {
    await producer.connect();
    await producer.send({
      topic: 'movie.audit',
      messages: [{
        value: JSON.stringify({
          event: `movie.${action}`,
          record_id: recordId,
          before,
          after,
          timestamp: new Date().toISOString()
        })
      }]
    });
  } catch (error) {
    console.error('Failed to publish audit event to Kafka:', error);
  } finally {
    await producer.disconnect();
  }
}
